network-area-diagram-viewer/src/main/resources/svg.js

```diff
@@ -0,0 +1,463 @@
+
+// TODO(Luma) consider three winding transformers
+// TODO(Luma) consider nodes with multiple buses
+// TODO(Luma) separate components of edges, having one that can be "scaled" (length line from exit of node to middle point)
+//              for reference, see Android NinePatch drawables
+//              https://developer.android.com/develop/ui/views/graphics/drawables#nine-patch
+// TODO(Luma) JavaFX seems to have problems drawing a translated foreign object, try to enclose it in a <g> element?
+// TODO(Luma) update on client: find related hidden nodes and update initial position in metadata?
+
+window.addEventListener('load', function() {
+    var svgDiagram = document.getElementsByTagName("svg")[0];
+    if (svgDiagram != null) {
+        makeDraggableSvg(svgDiagram);
+    }
+});
+
+// Draggable SVG Elements from
+// https://www.petercollingridge.co.uk/tutorials/svg/interactive/dragging/
+function makeDraggableSvg(svg) {
+    var selectedElement = false, offset, transform;
+    var diagram = new Diagram(svg);
+
+    svg.addEventListener('mousedown', startDrag);
+    svg.addEventListener('mousemove', drag);
+    svg.addEventListener('mouseup', endDrag);
+    svg.addEventListener('mouseleave', endDrag);
+
+    function getMousePosition(event) {
+        // From screen coordinates to SVG coordinate system
+        var CTM = svg.getScreenCTM();
+        return {
+            x: (event.clientX - CTM.e) / CTM.a,
+            y: (event.clientY - CTM.f) / CTM.d
+        };
+    }
+
+    function startDrag(event) {
+        var draggableElem = diagram.getDraggableFrom(event);
+        if (!draggableElem) {
+            return;
+        }
+        selectedElement = draggableElem;
+        selectedElement.style.cursor = 'grabbing';
+
+        offset = getMousePosition(event);
+        var transforms = getTransformsEnsuringFirstIsTranslation(svg, selectedElement);
+        // Get initial translation amount
+        transform = transforms.getItem(0);
+        offset.x -= transform.matrix.e;
+        offset.y -= transform.matrix.f;
+    }
+
+    const CONTINUOUS_UPDATE = true;
+    function drag(event) {
+        if (selectedElement) {
+            event.preventDefault();
+            var position = getMousePosition(event);
+            var translation = {x: position.x - offset.x, y: position.y - offset.y};
+            transform.setTranslate(translation.x, translation.y);
+            if (CONTINUOUS_UPDATE) {
+                diagram.update(selectedElement, position, translation);
+            }
+        }
+    }
+
+    function endDrag(event) {
+        if (selectedElement) {
```

Comment:
I think the metadata should be updated if not on the move, at least at the end

Reply:
Agree. Once the operation is completed, we should update the metadata (but we have to be aware that we also need the original metadata to know the transformations we have to apply to the elements)